ls1intum/Artemis
# `Programming exercises`: Improve structural solution entry generation

### Checklist
#### General


#### Server
#### Changes affecting Programming Exercises

### Motivation and Context


Follow up to #4760.
#4760 introduced the first version of the generation of solution entries for structural test cases. This first version did not have the best code quality and did not support annotations.

### Description

The code quality has been improved by utilizing the visitor pattern for source code generation and other more general improvements.
Annotations are now also supported for classes, methods, attributes, and constructors. Annotations for parameters are not supported, as this is something that is not part of the test.json.

This PR does not address the fact that there is no proper view for solution entries, as this is part of another upcoming PR. So the manual test still requires a look into the console.

### Steps for Testing

Prerequisites:
- 1 Admin
- 1 Programming Exercise

1. Log in to Artemis
2. Navigate to the Exercise Detail Page
3. Edit the solution repository (Add annotations)
4. Edit the test.json in the test repository (Add annotations)
5. Open the dev console
6. Click on the button "Create structural solution entries"
    - Please note that the changes might only be applied after the solution build plan is finished, as not all test cases might be present yet)
8. Check that the solution entries have been created properly (with annotations) according to the solution repo and test.json
    - Check classes
    - Check attributes
    - Check constructors
    - Check methods

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| StructuralAttribute.java | 87% | 100% |
| StructuralClass.java | 94% | 100% |
| StructuralClassElements.java | n/a | 100% |
| StructuralConstructor.java | 66% | 98% |
| StructuralElement.java | 64% | 87% |
| StructuralAttribute.java | 87% | 100% |
| StructuralMethod.java | 55% | 88% |
| StructuralTestCaseService.java | 91% | 88% |
